classes/class.tapestry-user-progress.php

```diff
@@ -218,12 +141,16 @@ private function _complete()
         update_user_meta($this->_userId, 'tapestry_'.$this->postId.'_node_completed_'.$this->nodeMetaId, true);
     }
 
-    private function _completeQuestion($questionId)
+    private function _completeQuestion($questionId, $answerType, $answerData)
     {
-        $nodeMetadata = get_metadata_by_mid('post', $this->nodeMetaId)->meta_value;
-        $quiz = $this->_getQuizProgress($this->nodeMetaId, $nodeMetadata, $this->_userId);
-        $quiz[$questionId]['completed'] = true;
-        update_user_meta($this->_userId, 'tapestry_'.$this->postId.'_node_quiz_'.$this->nodeMetaId, $quiz);
+        $userAnswer = get_user_meta($this->_userId, 'tapestry_'.$this->postId.'_'.$this->nodeMetaId.'_question_'.$questionId.'_answers', true);
+        if ($userAnswer === "") {
+            $newUserAnswer[$answerType] = $answerData;
+            update_user_meta($this->_userId, 'tapestry_'.$this->postId.'_'.$this->nodeMetaId.'_question_'.$questionId.'_answers', $newUserAnswer);
+        } else {
+            $userAnswer[$answerType] = $answerData;
+            update_user_meta($this->_userId, 'tapestry_'.$this->postId.'_'.$this->nodeMetaId.'_question_'.$questionId.'_answers', $userAnswer);
+        }
```

Comment:
@wynnset that's interesting. I didn't know you could do that. But I did some reading into this and the [official php documentation](https://www.php.net/manual/en/language.types.array.php) says that this way of initializing an array is discouraged as it could have some unwanted consequences.

Also, you can avoid your whole if/else statement altogether if you put `|| array()` at the end of line 146 so that if it's an empty string, [it evaluates to Boolean false](https://www.php.net/manual/en/language.types.boolean.php) and so would return `array()` as the value of `$userAnswer` and you can just keep the content of your else statement that way.
```suggestion
        $userAnswer = get_user_meta($this->_userId, 'tapestry_'.$this->postId.'_'.$this->nodeMetaId.'_question_'.$questionId.'_answers', true) || array();
        $userAnswer[$answerType] = $answerData;
        update_user_meta($this->_userId, 'tapestry_'.$this->postId.'_'.$this->nodeMetaId.'_question_'.$questionId.'_answers', $userAnswer);
```

Reply:
Hey Aidin, can I refactor it to this way instead? It seems that get_user_meta($this->_userId, 'tapestry_'.$this->postId.'_'.$this->nodeMetaId.'_question_'.$questionId.'_answers', true) || array() is returning a boolean value of true which then assigns 1 to $userAnswer. 

$userAnswer = get_user_meta($this->_userId, 'tapestry_'.$this->postId.'_'.$this->nodeMetaId.'_question_'.$questionId.'_answers', true);
         if ($userAnswer === "") {
             $userAnswer= array();
         }
        $userAnswer[$answerType] = $answerData;
        update_user_meta($this->_userId, 'tapestry_'.$this->postId.'_'.$this->nodeMetaId.'_question_'.$questionId.'_answers', $userAnswer);

Reply:
@harry02031 yes that works. Sorry I thought that notation works, but it doesn't.